src/Algorithms/HungarianAlgorithm.cs

```diff
@@ -0,0 +1,387 @@
+﻿using System;
+using OpenTK;
+
+namespace Bearded.Utilities.Algorithms
+{
+    /// <summary>
+    /// The Hungarian algorithm is a cubic algorithm that solves the assignment problem.
+    /// The algorithm calculates a maximum matching between workers and jobs, while minimising the total cost.
+    /// </summary>
+    /// <remarks>The original implementation was by Kevin L. Stern.</remarks>
+    public sealed class HungarianAlgorithm
+    {
+        #region Public static interface
+        /// <summary>
+        /// Evaluates a minimal cost matching using the given cost matrix.
+        /// </summary>
+        /// <param name="costMatrix">A matrix of costs of assigning jobs to workers.</param>
+        /// <returns>The minimal cost matching based on the specified cost matrix. A value of -1 means the worker is not assigned.</returns>
+        public static int[] Run(float[,] costMatrix)
+        {
+            var instance = new HungarianAlgorithm(costMatrix);
+            return instance.execute();
+        }
+
+        /// <summary>
+        /// Evaluates a minimal cost matching using the given metric.
+        /// </summary>
+        /// <typeparam name="TWorker">The type representing the workers.</typeparam>
+        /// <typeparam name="TJob">The type representing the jobs.</typeparam>
+        /// <param name="workers">The elements representing the workers.</param>
+        /// <param name="jobs">The elements representing the jobs.</param>
+        /// <param name="getCost">A function that calculates the cost of assigning a job to a worker.</param>
+        /// <returns>The minimal cost matching based on the specified metric. A value of -1 means the worker is not assigned.</returns>
+        public static int[] Run<TWorker, TJob>(TWorker[] workers, TJob[] jobs, Func<TWorker, TJob, float> getCost)
+        {
+            var costMatrix = new float[workers.Length, jobs.Length];
+            for (int t = 0; t < jobs.Length; t++)
+                for (int s = 0; s < workers.Length; s++)
+                    costMatrix[s, t] = getCost(workers[s], jobs[t]);
+
+            return HungarianAlgorithm.Run(costMatrix);
+        }
+
+        /// <summary>
+        /// Evaluates a minimal cost matching on two-dimensional vectors using the least-squares metric.
+        /// </summary>
+        /// <param name="from">The source vectors.</param>
+        /// <param name="to">The destination vectors.</param>
+        /// <returns>The minimal cost matching based on the least-squares metric. A value of -1 means the source vector is not assigned.</returns>
+        public static int[] Run(Vector2[] from, Vector2[] to)
+        {
+            return HungarianAlgorithm.Run(from, to, (u, v) => (v - u).LengthSquared);
```

Comment:
I'd be curious to see how much overhead this lambda adds, compared to writing out the loop directly.

Since this will likely be one of the most common usages of the algorithm, it may be worth testing this, and possibly writing out the loop from above.